portal-impl/src/com/liferay/portal/util/HtmlImpl.java

```diff
@@ -225,50 +225,65 @@ else if (mode == ESCAPE_MODE_URL) {
 
 		StringBuilder sb = new StringBuilder(text.length());
 
+		char[] hexBuffer = new char[4];
+
+		boolean modified = false;
+
 		for (int i = 0; i < text.length(); i++) {
 			char c = text.charAt(i);
 
-			if ((mode == ESCAPE_MODE_ATTRIBUTE) &&
-				(!_isValidXmlCharacter(c) ||
-				 _isUnicodeCompatibilityCharacter(c))) {
+			if (c < 256) {
+				if ((c < 128) && _validChars[c]) {
+					sb.append(c);
+				}
+				else if ((mode == ESCAPE_MODE_ATTRIBUTE) &&
```

Comment:
This may change the meaning of the text. In the specific case of the text being a piece of code, this makes really difficult to understand escaped code, as formatting is lost.